### PR TITLE
feat: add c++ api

### PR DESCRIPTION
**Pull request by ddiakiteaneo**
This is the armoniK API project c++ project. The PR contains:

- Project Cmake files 
- Worker server
- Submitter client methods 
- CI pipeline for build and code formatting
- Unit and mock test of the client submitter using gtest

### CODE REVIEW

File: .dockerignore
Comment (aneojgurhem) [inline .dockerignore:1]: you should take a look at https://github.com/github/gitignore/blob/master/VisualStudio.gitignore
and the .dockerignore in core

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,8 +1,9 @@
 name: CI
 
 on:
-  pull_request:
-
+  push:
```
Comment (esoubiran-aneo) [inline .github/workflows/ci.yml:4]: Please revert to "pull_request"

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -76,6 +77,51 @@ jobs:
...
+    timeout-minutes: 15
+    
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+        with:
+          ref: ${{ github.ref }}
+          fetch-depth: 0
+          submodules: true
```
Comment (esoubiran-aneo) [inline .github/workflows/ci.yml:92]: Is there submodule in this package?

File: .github/workflows/publish-edge.yml [hunk 3]
```diff
@@ -8,6 +8,7 @@ on:
     branches:
       - main
 
+
```
Comment (esoubiran-aneo) [inline .github/workflows/publish-edge.yml:11]: Please, remove this extra line

File: packages/cpp/README.md
Comment (esoubiran-aneo) [inline packages/cpp/README.md:1]: Please, consider using `.docs` for you documentation and not a README.

File: packages/cpp/tools/RefreshEnv.cmd
Comment (esoubiran-aneo) [inline packages/cpp/tools/RefreshEnv.cmd:1]: Do not understand what this file does

File: packages/cpp/tools/build-worker.sh [hunk 4]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+
+# Declare a variable that will store the image tag for our Docker image
+IMAGE_TAG="armonik-api-cpp:v0.1"
```
Comment (esoubiran-aneo) [inline packages/cpp/tools/build-worker.sh:4]: Image version is always 0.1?

**Review by esoubiran-aneo — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 1 above)
Comment (ddiakiteaneo) [reply .github/workflows/ci.yml]: Done !

File: .github/workflows/ci.yml (hunk 2 above)
Comment (ddiakiteaneo) [reply .github/workflows/ci.yml]: No, I removed the line !

File: .github/workflows/publish-edge.yml (hunk 3 above)
Comment (ddiakiteaneo) [reply .github/workflows/publish-edge.yml]: Done !

File: packages/cpp/tools/RefreshEnv.cmd
Comment (ddiakiteaneo) [reply packages/cpp/tools/RefreshEnv.cmd:1]: This file reads the environment variables from the registry and sets the session environment variables on Windows for compilation. The file is used in the bootstrap-win64.ps1 file during the project compilation on Windows.

File: packages/cpp/tools/build-worker.sh (hunk 4 above)
Comment (ddiakiteaneo) [reply packages/cpp/tools/build-worker.sh:4]: No, it will be variabilize!

File: packages/cpp/ArmoniK.Api.Tests/source/SubmitterCLientTest.cpp [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+  SubmitterClient submitter(&stub);
+  std::string session_id =
+      submitter.create_session(task_options, partition_ids);
+
+  std::cout << "create_session response: " << session_id << std::endl;
+
+  ASSERT_TRUE(session_id.empty());
+}
+
+TEST(testMock, cancleSession) {
```
Comment (lemaitre-aneo) [inline packages/cpp/ArmoniK.Api.Tests/source/SubmitterCLientTest.cpp:71]:
```suggestion
TEST(testMock, cancelSession) {
```

File: packages/cpp/ArmoniK.Api.Tests/source/SubmitterCLientTest.cpp [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+  task_options.mutable_max_duration()->set_seconds(3600);
+  task_options.mutable_max_duration()->set_nanos(0);
+  task_options.set_max_retries(3);
+  task_options.set_priority(1);
+  task_options.set_partition_id("cpp");
+  task_options.set_application_name("my-app");
+  task_options.set_application_version("1.0");
+  task_options.set_application_namespace("my-namespace");
+  task_options.set_application_service("my-service");
+  task_options.set_engine_type("Unified");
```
Comment (lemaitre-aneo) [inline packages/cpp/ArmoniK.Api.Tests/source/SubmitterCLientTest.cpp:96]: This is duplicated. You could most likely create a function (or static method) to create such a task options.

File: packages/cpp/ArmoniK.Api.Worker/header/utils/WorkerServer.h [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+ * @class WorkerServer
+ * @brief Represents the worker server for ArmoniK API.
+ */
+class WorkerServer {
+ public:
+  common::serilog::serilog logger;
+
+ private:
+  ::grpc::ServerBuilder builder_;
+  std::shared_ptr<common::utils::IConfiguration> configuration_;
```
Comment (lemaitre-aneo) [inline packages/cpp/ArmoniK.Api.Worker/header/utils/WorkerServer.h:34]: I would say it is not necessary to store the configuration and the builder during the lifetime of the server.

File: packages/cpp/tools/RefreshEnv.cmd
Comment (lemaitre-aneo) [inline packages/cpp/tools/RefreshEnv.cmd:1]: What is the use of this script?

File: packages/cpp/tools/Dockerfile.ubuntu
Comment (lemaitre-aneo) [inline packages/cpp/tools/Dockerfile.ubuntu:1]: What is the use of this docker image?

File: packages/cpp/tools/Dockerfile.ubuntu [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+  && rm -rf /var/lib/apt/lists/* \
+  # Download and install CMake
+  && wget https://github.com/Kitware/CMake/releases/download/v3.24.1/cmake-3.24.1-Linux-x86_64.sh \
+      -q -O /tmp/cmake-install.sh \
+      && chmod u+x /tmp/cmake-install.sh \
+      && mkdir /opt/cmake-3.24.1 \
+      && /tmp/cmake-install.sh --skip-license --prefix=/opt/cmake-3.24.1 \
+      && rm /tmp/cmake-install.sh \
+      # Create symlinks for the CMake binary
+      && ln -s /opt/cmake-3.24.1/bin/* /usr/local/bin
```
Comment (lemaitre-aneo) [inline packages/cpp/tools/Dockerfile.ubuntu:31]:
```suggestion
RUN apt-get update && \
  apt-get install build-essential wget cmake
```

File: packages/cpp/tools/Dockerfile.worker [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    make install && \
+    ln -sf /app/install/lib/libgrpc++.so /app/install/lib/libgrpc++.so.1 && \
+    ln -sf /app/install/lib/libgrpc++_reflection.so /app/install/lib/libgrpc++_reflection.so.1
+
+# Update the PATH environment variable to include the gRPC libraries and binaries
+ENV PATH="/app/install/lib:$PATH"
+ENV PATH="/app/install/bin:$PATH"
+
+# Remove the gRPC source folder to save space in the final image
+RUN rm -rf /tmp/grpc
```
Comment (lemaitre-aneo) [inline packages/cpp/tools/Dockerfile.worker:46]: This does not work the way you think. New layers can only *increase* the image size, never decrease.

File: packages/cpp/tools/bootstrap-win64.ps1
Comment (lemaitre-aneo) [inline packages/cpp/tools/bootstrap-win64.ps1:1]: What is the use for this script? Is the development environment so complex to setup that we need a script???

File: packages/cpp/tools/build-worker.sh
Comment (lemaitre-aneo) [inline packages/cpp/tools/build-worker.sh:1]: Aren't we reinventing the wheel for build scripts here?

File: packages/cpp/tools/patch/0001-Fix-issue-with-Visual-Studio-2022-toolset.patch
Comment (lemaitre-aneo) [inline packages/cpp/tools/patch/0001-Fix-issue-with-Visual-Studio-2022-toolset.patch:1]: Is this really necessary? If so, where it is documented?

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -2,7 +2,7 @@ name: CI
 
 on:
   pull_request:
-
+
```
Comment (esoubiran-aneo) [inline .github/workflows/ci.yml:5]: There is extra space here that can be removed.

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -76,6 +76,50 @@ jobs:
             exit 1
           fi
 
+  format-cpp:
+    name: Format C++
+    runs-on: ubuntu-latest
+    timeout-minutes: 15
+
```
Comment (esoubiran-aneo) [inline .github/workflows/ci.yml:83]: Useless extra line

File: .github/workflows/publish-edge.yml [hunk 12]
```diff
@@ -224,3 +224,27 @@ jobs:
...
+      - name: Checkout
+        uses: actions/checkout@v3
+        with:
+          ref: ${{ github.ref }}
+
+      - name: Build the package
+        run: |
+          cd packages/cpp/tools/
+          ./compile.sh
+
```
Comment (esoubiran-aneo) [inline .github/workflows/publish-edge.yml:245]: Remove extra line

File: .github/workflows/publish-edge.yml
Comment (esoubiran-aneo) [inline .github/workflows/publish-edge.yml:1]: You muse merge main into your branch in order to support new way for version creation

**Review by esoubiran-aneo — CHANGES_REQUESTED** [review]

**Review by lemaitre-aneo — CHANGES_REQUESTED** [review]
I think there is a missing clang format configuration file. Could it work with the `.editorconfig`? If so, it would be nice to have a proper configuration for c++

File: packages/cpp/ArmoniK.Api.Tests/source/SubmitterCLientTest.cpp [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+  ClientContext context;
+  CreateSessionReply reply;
+  CreateSessionRequest request;
+
+  const std::vector<std::string>& partition_ids = {"default"};
+
+  TaskOptions task_options;
+
+  std::shared_ptr<Channel> channel = grpc::CreateChannel(
+      "172.30.39.223:5001", grpc::InsecureChannelCredentials());
```
Comment (dbrasseur-aneo) [inline packages/cpp/ArmoniK.Api.Tests/source/SubmitterCLientTest.cpp:43]: I know it's a local address, but maybe find a way to remove it here ?

File: packages/cpp/tools/patch/0001-Fix-issue-with-Visual-Studio-2022-toolset.patch [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+ #elif defined(_MSC_VER) && !defined(__clang__)
+ #define alignas(x) __declspec(align(x))
+ #define alignof __alignof
++#endif
+ #else
+ // With the exception of MSVC, we require C11 to build the library. C11 is a
+ // prerequisite for improved refcounting performance. All our supported C
+-- 
+2.25.1
+
```
Comment (dbrasseur-aneo) [inline packages/cpp/tools/patch/0001-Fix-issue-with-Visual-Studio-2022-toolset.patch:30]: Not sure this belongs in the code

**Review by dbrasseur-aneo — CHANGES_REQUESTED** [review]

File: packages/cpp/ArmoniK.Api.Client/header/submitter/SubmitterClient.h [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+  // SubmitterClient(std::unique_ptr<armonik::api::grpc::v1::submitter::Submitter::StubInterface>
+  // stub);
+
+  /**
+   * @brief Construct a new Submitter Client:: Submitter Client object
+   *
+   * @param stub the gRPC client stub
+   */
+  SubmitterClient(
+      armonik::api::grpc::v1::submitter::Submitter::StubInterface* stub);
```
[the remark this answers is not in the document]
Comment (ddiakiteaneo) [reply packages/cpp/ArmoniK.Api.Client/header/submitter/SubmitterClient.h]: It was for test purposes, I replaced it with a smart ptr.

File: packages/cpp/ArmoniK.Api.Client/header/submitter/SubmitterClient.h [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+
+  /**
+   * @brief Initializes task options creates channel with server address
+   *
+   * @param channel The gRPC channel to communicate with the server.
+   * @param task_options The task options.
+   */
+  static void init(std::shared_ptr<grpc::Channel>& channel,
+                   armonik::api::grpc::v1::TaskOptions& task_options);
```
[the remark this answers is not in the document]
Comment (ddiakiteaneo) [reply packages/cpp/ArmoniK.Api.Client/header/submitter/SubmitterClient.h]: Removed!

File: packages/cpp/ArmoniK.Api.Client/CMakeLists.txt [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+
+file(GLOB PROTO_FILES 
+"${PROTO_FILES_DIR}/submitter_service.proto"
+"${PROTO_FILES_DIR}/applications_service.proto"
+"${PROTO_FILES_DIR}/sessions_service.proto"
+"${PROTO_FILES_DIR}/tasks_service.proto"
+"${PROTO_FILES_DIR}/results_service.proto"
+"${PROTO_FILES_DIR}/auth_service.proto"
+"${PROTO_FILES_DIR}/partitions_service.proto"
+"${PROTO_FILES_DIR}/events_service.proto")
```
[the remark this answers is not in the document]
Comment (ddiakiteaneo) [reply packages/cpp/ArmoniK.Api.Client/CMakeLists.txt]: For Windows portability concerns, I did not use the script.

File: .docs/content/1.guide/5.cpp.md [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+- CMake
+- NASM
+
+## Compilation Steps for windows
+Follow these steps to compile the Armonik API client and server:
+
+- From a powershell, go to the folder package/cpp/tools 
+
+```powershell
+cd packages\cpp/tools
```
Comment (ngruelaneo) [inline .docs/content/1.guide/5.cpp.md:56]:
```suggestion
cd packages\cpp\tools

```

File: packages/cpp/ArmoniK.Api.Client/CMakeLists.txt [hunk 19]
```diff
@@ -0,0 +1,116 @@
+set(PROJECT_NAME Armonik.Api.Client)
+set(NAMESPACE Armonik::Api::Client)
+set(version 0.1.0)
```
Comment (ngruelaneo) [inline packages/cpp/ArmoniK.Api.Client/CMakeLists.txt:3]: Why does this version is not the same than API?

File: packages/cpp/tools/Dockerfile.worker (hunk 9 above)
Comment (ngruelaneo) [reply packages/cpp/tools/Dockerfile.worker:46]: To add more information: RUN, COPY and ADD docker functions are creating new layer. Here you probably want to move this line to line 40 (without the RUN)

File: packages/cpp/tools/RefreshEnv.cmd [hunk 20]
```diff
@@ -0,0 +1,67 @@
+@echo off
```
Comment (ngruelaneo) [inline packages/cpp/tools/RefreshEnv.cmd:1]: This file cannot be in this PR or the license should be respected.

**Review by ngruelaneo — CHANGES_REQUESTED** [review]

**Pull request closed without merging**